### PR TITLE
fix nil function invocation in client-go/tools/leaderelection

### PR DESCRIPTION
**Pull request by hbagdi**
It should be possible for the user of leaderelection API to not provide
callbacks for stopped and started leading conditions. Currnetly, the
code results in fatal error and exits.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by hbagdi** [comment]
I signed it

**Review by neolit123 — COMMENTED** [review]
thank you for the PR, i think this make sense, but needs approval by the maintainers.
the 1.15 code freeze will be lifted after 11th of june.
try pinging #sig-api-machinery on slack if needed after that.

/ok-to-test
/priority backlog

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go [hunk 1]
```diff
@@ -190,14 +190,18 @@ type LeaderElector struct {
 func (le *LeaderElector) Run(ctx context.Context) {
 	defer func() {
 		runtime.HandleCrash()
-		le.config.Callbacks.OnStoppedLeading()
+		if le.config.Callbacks.OnStoppedLeading != nil {
+			go le.config.Callbacks.OnStoppedLeading()
```
Comment (cheftako) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Probably not a good idea to switch this to a new goroutine without serious investigation. Our general behavior in things like the KCM to losing leader election is to kill our own process. Moving this to a goroutine means that the callback is now racing to act prior to the process being killed.

Comment (hbagdi) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: That makes sense. I re-uped the commit and removed the goroutine.

**Review by cheftako — COMMENTED** [review]
/lgtm

**Comment by roycaihw** [comment]
looks good

/assign @mikedanese
for approval

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go [hunk 2]
```diff
@@ -190,14 +190,18 @@ type LeaderElector struct {
...
+			le.config.Callbacks.OnStoppedLeading()
+		}
 	}()
 	if !le.acquire(ctx) {
 		return // ctx signalled done
 	}
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	go le.config.Callbacks.OnStartedLeading(ctx)
+	if le.config.Callbacks.OnStartedLeading != nil {
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: The other one is fine, but if this is nil, we should probably panic. There's no point to this client if you don't do anything when you get leadership.

Comment (hbagdi) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: A client could potentially not actively want to be notified of it being a leader and but choose to check the leader status using `IsLeader()` function whenever it is actually going to take up any action.

Comment (mikedanese) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: You can do this by providing an empty OnStartedLeading callback. I am hesitant to further facilitate this behavior.

Comment (cheftako) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Mike has a good point. That having been said I personally would rather an explicit panic which explained the issue (eg. "Leader elect called without a proper started leading method") than a fatal error about referencing invalid memory.

Comment (hbagdi) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: >  You can do this by providing an empty OnStartedLeading callback. I am hesitant to further facilitate this behavior.

My take is that zero values in Go shouldn't panic in such a case.

Please let me know the exact change that should be made to move this forward. You guys are more familiar with the API style and end-user expectations here so I'll make a change that works for everyone.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go [hunk 3]
```diff
@@ -190,14 +190,18 @@ type LeaderElector struct {
 func (le *LeaderElector) Run(ctx context.Context) {
 	defer func() {
 		runtime.HandleCrash()
-		le.config.Callbacks.OnStoppedLeading()
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Sorry I misread this originally. This is intended to crash.

**Comment by mikedanese** [comment]
Can you validate that OnStart and OnFinish callbacks are non-nil in NewLeaderElector, else return an error?

**Comment by hbagdi** [comment]
@mikedanese updated. Please re-review.

**Comment by hbagdi** [comment]
How do we fix the failing CI? 🤔

**Pull request merged**